### PR TITLE
Improve ChibiOS memory functions

### PR DESCRIPTION
**Pull request by networkfusion**
## Description
This PR provides the platform specific implementation of memory (free, malloc, realloc) functions for ChibiOS.

## Motivation and Context
We currently just revert to the generic functions which may cause issues since other functions within the RTOS are using the heap in a different way.

If this is not required, we should seriously reconsider the use of `platform_heap` as all targets currently use the generic (default) implementation.

## How Has This Been Tested?
Tested ability to run locally on an STM32F769I
May need further tests to ensure full compatibility (although please give examples of what needs testing).

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dev Containers (changes related with Dev Containers, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist

- [ ] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [ ] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [ ] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).

### CODE REVIEW

**Comment by josesimoes** [comment]
I was reviewing this code and just realized that it's reverting a global rework for platform malloc/free done in #2076 
Back then ChibiOS API was being used and it was changed to the CRT malloc API because to make it easy to better deal with the memory allocation on each target.

Reverting this requires testing on all STM32 targets.

File: targets/ChibiOS/_common/platform_heap.c [hunk 1]
```diff
@@ -4,40 +4,63 @@
...
+    {
+        return chHeapAlloc(NULL, size);
+    }
+
+    /* previous allocated segment is preceded by an heap_header */
+    hp = ptr - sizeof(union heap_header);
+    prev_size = hp->used.size; /* size is always multiple of 8 */
+
+    /* check new size memory alignment */
+    if (size % 8 == 0)
```
Comment (josesimoes) [inline targets/ChibiOS/_common/platform_heap.c:41]: Why the 8 bytes aligment?

File: targets/ChibiOS/_common/platform_heap.c [hunk 2]
```diff
@@ -4,40 +4,63 @@
...
-
-// define back
-#define free YOU_SHALL_NOT_USE_free
+    if (ptr)
+    {
+        chHeapFree(ptr);
+    }
 }
 
 void *platform_realloc(void *ptr, size_t size)
```
Comment (josesimoes) [inline targets/ChibiOS/_common/platform_heap.c:24]: We're better removing realloc all toghether. This is not used anywhere. No good reason to add unused code... 
If you're doing it, please make it on a separate PR as this will touch all platforms.

Comment (networkfusion) [reply targets/ChibiOS/_common/platform_heap.c]: Now removed inline with main.

File: targets/ChibiOS/_common/platform_heap.c (hunk 1 above)
Comment (networkfusion) [reply targets/ChibiOS/_common/platform_heap.c]: No longer needed.

**Comment by josesimoes** [comment]
Switching to CH heap requires changing the memory allocation config in all SMT32 targets and retest everything, specially the network and TLS stuff. 😱 
It's working nicely as it is now. Don't think it's worth the effort to mess with it.

**Comment by networkfusion** [comment]
Closing as `fixed`. It is still available in the history if required.
